File: 145 HackerRank.java
// 33% HackerRank

import java.util.*;

public class Main {
    public static boolean isReversible(long n) {
        String s = Long.toString(n);
        if (s.charAt(s.length() - 1) == '0') {
            return false;
        }
        long reversedN = Long.parseLong(new StringBuilder(s).reverse().toString());
        long total = n + reversedN;
        String totalStr = Long.toString(total);
        for (int i = 0; i < totalStr.length(); i++) {
            if ((totalStr.charAt(i) - '0') % 2 == 0) {
                return false;
            }
        }
        return true;
    }

    public static List<Long> precomputeReversibleNumbers(long limit) {
        List<Long> reversibleNumbers = new ArrayList<>();
        for (long n = 1; n < limit; n++) {
            if (isReversible(n)) {
                reversibleNumbers.add(n);
            }
        }
        return reversibleNumbers;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int T = scanner.nextInt();
        long[] queries = new long[T];
        long maxQuery = 0;
        for (int i = 0; i < T; i++) {
            queries[i] = scanner.nextLong();
            if (queries[i] > maxQuery) {
                maxQuery = queries[i];
            }
        }
        List<Long> reversibleNumbers = precomputeReversibleNumbers(maxQuery + 1);
        Collections.sort(reversibleNumbers);
        for (long N : queries) {
            int left = 0;
            int right = reversibleNumbers.size();
            while (left < right) {
                int mid = (left + right) / 2;
                if (reversibleNumbers.get(mid) < N) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
            System.out.println(left);
        }
    }
}

File: 145 HackerRank.py
# 33% HackerRank

def is_reversible(n):
    s = str(n)
    if s[-1] == '0':
        return False
    reversed_n = int(s[::-1])
    total = n+reversed_n
    for digit in str(total):
        if int(digit) % 2 == 0:
            return False
    return True

def precompute_reversible_numbers(limit):
    reversible = []
    for n in range(1, limit):
        if is_reversible(n):
            reversible.append(n)
    return reversible

T = int(input())
queries = []
for _ in range(T):
    queries.append(int(input()))
max_query = max(queries)
reversible_numbers = precompute_reversible_numbers(max_query + 1)
reversible_numbers.sort()
for N in queries:
    left, right = 0, len(reversible_numbers)
    while left < right:
        mid = (left + right) // 2
        if reversible_numbers[mid] < N:
            left = mid + 1
        else:
            right = mid
    print(left)

